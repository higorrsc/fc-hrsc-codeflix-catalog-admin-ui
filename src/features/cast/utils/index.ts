import { CastMemberType } from '../types';

export function getCastMemberTypeOptions() {
  return Object.values(CastMemberType)
    .filter((value) => typeof value === 'number')
    .map((value) => ({
      value,
      label:
        CastMemberType[value as CastMemberType].charAt(0) +
        CastMemberType[value as CastMemberType].slice(1).toLowerCase(),
    }));
}
