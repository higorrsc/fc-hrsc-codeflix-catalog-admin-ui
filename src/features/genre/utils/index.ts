import { Category } from '../../category/types';
import { Genre } from '../types';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
