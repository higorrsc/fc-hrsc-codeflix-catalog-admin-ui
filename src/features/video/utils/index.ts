import { Video, VideoPayload } from 'src/features/video/types';

export const mapVideoToForm = (video: Video): VideoPayload => {
  return {
    id: video.id,
    title: video.title,
    description: video.description,
    year_launched: video.year_launched,
    opened: video.opened,
    rating: video.rating,
    duration: video.duration,
    genres_id: video.genres?.map((genre) => genre.id),
    categories_id: video.categories?.map((category) => category.id),
    cast_members_id: video.cast_members?.map((cast_member) => cast_member.id),
  };
};
