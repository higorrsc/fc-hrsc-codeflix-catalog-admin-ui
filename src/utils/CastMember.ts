import { CastMemberType } from '../types/CastMember';

const castMemberTypeLabelMap: Record<CastMemberType, string> = {
  [CastMemberType.ACTOR]: 'Actor',
  [CastMemberType.DIRECTOR]: 'Director',
};

export function getCastMemberTypeLabel(type: CastMemberType): string {
  return castMemberTypeLabelMap[type] || 'Unknown';
}

export const castMemberTypeOptions = [
  { value: CastMemberType.ACTOR, label: 'Actor' },
  { value: CastMemberType.DIRECTOR, label: 'Director' },
];
