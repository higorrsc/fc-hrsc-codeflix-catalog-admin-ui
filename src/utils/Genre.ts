import { Category } from '../types/Category';
import { Genre } from '../types/Genre';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
