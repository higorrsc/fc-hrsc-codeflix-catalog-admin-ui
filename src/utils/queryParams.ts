import { ListOptions } from '../types/_shared'; // Import ListOptions if needed for typing map

/**
 * Maps specific frontend parameter names (keys often matching ListOptions)
 * to their corresponding backend query parameter names expected by the API.
 * We use Partial<Record<keyof ListOptions | string, string>> to allow mapping
 * ListOptions keys and potentially other string keys if needed later.
 */
const paramKeyMap: Partial<Record<keyof ListOptions | string, string>> = {
  perPage: 'per_page',
  // Add other mappings if needed
};

/**
 * Parses an object containing potential query parameters into a URL query string
 * suitable for API requests.
 *
 * - Accepts an object like `ListOptions` or any object with potential query params.
 * - Filters out parameters with `null` or `undefined` values.
 * - Explicitly ignores UI-specific keys like `rowsPerPage`.
 * - Maps keys according to `paramKeyMap` (e.g., 'perPage' becomes 'per_page').
 * - Converts remaining parameter values to strings.
 *
 * @template T - The type of the params object (e.g., ListOptions).
 * @param params - An object containing potential query parameters.
 * @returns A URL query string (without the leading '?'). Returns an empty string
 *          if no valid parameters are found after filtering and mapping.
 */
export function parseQueryParams<T extends Record<string, any>>(
  params: T
): string {
  const queryParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    // 1. Skip null/undefined/empty string for search
    if (
      value === null ||
      value === undefined ||
      (key === 'search' && value === '') // Also skip empty search strings
    ) {
      return;
    }

    // 2. Skip known UI-specific keys
    if (key === 'rowsPerPage') {
      return;
    }

    // 3. Determine the correct key name for the API
    const mappedKey = paramKeyMap[key as keyof typeof paramKeyMap] || key;

    // 4. Convert the value to a string and append
    queryParams.append(mappedKey, String(value));
  });

  return queryParams.toString();
}
