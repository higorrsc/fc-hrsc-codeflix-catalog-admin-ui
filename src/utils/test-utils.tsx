import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { ThemeProvider } from '@mui/system';
import { configureStore } from '@reduxjs/toolkit';
import { SnackbarProvider } from 'notistack';
import { BrowserRouter } from 'react-router-dom';
import type { RootState } from '../app/store';
import { appTheme } from '../config/theme';
import { apiSlice } from '../features/api/apiSlice';
import { castMembersApiSlice } from '../features/cast/castMemberSlice';
import { categoriesApiSlice } from '../features/categories/categorySlice';
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: ReturnType<typeof configureStore>;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        [categoriesApiSlice.reducerPath]: apiSlice.reducer,
        // @ts-ignore
        [castMembersApiSlice.reducerPath]: apiSlice.reducer,
      },
      // Adding the api middleware enables caching, invalidation, polling,
      // and other useful features of `rtk-query`.
      middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(apiSlice.middleware),
    }),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <React.StrictMode>
      <BrowserRouter>
        <Provider store={store}>
          <ThemeProvider theme={appTheme}>
            <SnackbarProvider maxSnack={3}>{children}</SnackbarProvider>
          </ThemeProvider>
        </Provider>
      </BrowserRouter>
    </React.StrictMode>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
