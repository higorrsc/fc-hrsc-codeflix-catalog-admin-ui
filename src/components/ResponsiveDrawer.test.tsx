import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render, screen, within } from '../utils/test-utils';
import { ResponsiveDrawer } from './ResponsiveDrawer';

const routes = [
  { path: '/categories', name: 'Categories' },
  { path: '/cast-members', name: 'Cast Members' },
  { path: '/genres', name: 'Genres' },
  { path: '/videos', name: 'Videos' },
];

describe('Test ResponsiveDrawer Component', () => {
  const mockOnClose = jest.fn();

  beforeEach(() => {
    mockOnClose.mockClear();
  });

  it('should render common drawer content (title and links) and match snapshot when open', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ResponsiveDrawer open={true} onClose={mockOnClose} />
      </MemoryRouter>
    );

    // Title - will be present in both temporary (dialog) and permanent drawers' DOM structure
    expect(screen.getAllByText('Codeflix').length).toBeGreaterThanOrEqual(1);
    // Specifically check within the dialog (temporary drawer)
    const temporaryDrawerWhenOpen = screen.getByTestId('drawer');
    expect(
      within(temporaryDrawerWhenOpen).getByText('Codeflix')
    ).toBeInTheDocument();

    // Links - check in dialog
    routes.forEach((route) => {
      const linkElement = within(temporaryDrawerWhenOpen).getByRole('link', {
        name: route.name,
      });
      expect(linkElement).toBeInTheDocument();
      expect(linkElement).toHaveAttribute('href', route.path);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call onClose when a link is clicked in the temporary drawer', () => {
    render(
      <MemoryRouter>
        <ResponsiveDrawer open={true} onClose={mockOnClose} />
      </MemoryRouter>
    );

    const temporaryDrawer = screen.getByTestId('drawer');
    const firstLink = within(temporaryDrawer).getByRole('link', {
      name: routes[0].name,
    });

    fireEvent.click(firstLink);
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  describe('Temporary Drawer (Mobile View)', () => {
    it('should be visible when open prop is true', () => {
      render(
        <MemoryRouter>
          <ResponsiveDrawer open={true} onClose={mockOnClose} />
        </MemoryRouter>
      );
      const drawerElement = screen.getByTestId('drawer');
      expect(drawerElement).toBeInTheDocument();
      expect(drawerElement).toBeVisible();

      // Check content within the temporary drawer
      expect(within(drawerElement).getByText('Codeflix')).toBeInTheDocument();
      routes.forEach((route) => {
        expect(
          within(drawerElement).getByRole('link', { name: route.name })
        ).toBeInTheDocument();
      });
    });

    it('should not be visible when open prop is false', () => {
      render(
        <MemoryRouter>
          <ResponsiveDrawer open={false} onClose={mockOnClose} />
        </MemoryRouter>
      );
      // Due to ModalProps={{ keepMounted: true }}, it might be in DOM but not visible.
      // The element with data-testid='drawer' will be in the document.
      const drawerElement = screen.getByTestId('drawer');
      expect(drawerElement).toBeInTheDocument();
      expect(drawerElement).not.toBeVisible();
    });

    it('should call onClose when Escape key is pressed', () => {
      render(
        <MemoryRouter>
          <ResponsiveDrawer open={true} onClose={mockOnClose} />
        </MemoryRouter>
      );
      const drawerElement = screen.getByTestId('drawer');

      fireEvent.keyDown(drawerElement, { key: 'Escape', code: 'Escape' });
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('Permanent Drawer (Desktop View)', () => {
    it('should always render its content structure', () => {
      // Render with temporary drawer closed to isolate permanent drawer's content
      render(
        <MemoryRouter>
          <ResponsiveDrawer open={false} onClose={mockOnClose} />
        </MemoryRouter>
      );

      // The permanent drawer is not a 'dialog'. Its content should be directly accessible.
      // We expect "Codeflix" titles.
      // Note: This assertion might be tricky if JSDOM renders both structures regardless of CSS.
      // We rely on the fact that when open=false, the temporary drawer (identified by data-testid='drawer') is not visible.
      // The permanent drawer's content should still be found.
      const titles = screen.getAllByText('Codeflix');
      // Depending on MUI's rendering with keepMounted, there might be a hidden one.
      // Let's find one that is not the temporary drawer.
      const permanentTitle = titles.find(
        (t) => t.closest('[data-testid="drawer"]') === null
      );
      expect(permanentTitle).toBeInTheDocument();

      routes.forEach((route) => {
        const links = screen.getAllByRole('link', { name: route.name });
        const permanentLink = links.find(
          (l) => l.closest('[data-testid="drawer"]') === null
        );
        expect(permanentLink).toBeInTheDocument();
        expect(permanentLink).toHaveAttribute('href', route.path);
      });
    });
  });
});
