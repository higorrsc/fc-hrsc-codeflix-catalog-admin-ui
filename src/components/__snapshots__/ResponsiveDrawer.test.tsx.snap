// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ResponsiveDrawer Component should render common drawer content (title and links) and match snapshot when open 1`] = `
<DocumentFragment>
  <nav
    class="MuiBox-root css-1o5pq7n"
  >
    <div
      class="MuiDrawer-root MuiDrawer-anchorLeft MuiDrawer-docked css-1bn13uf-MuiDrawer-docked"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-1lwhjos-MuiPaper-root-MuiDrawer-paper"
        style="--Paper-shadow: none;"
      >
        <div>
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-5lnhk4-MuiTypography-root"
            >
              Codeflix
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <a
              href="/categories"
              style="text-decoration: none;"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Categories
                    </span>
                  </div>
                </div>
              </li>
            </a>
            <a
              href="/cast-members"
              style="text-decoration: none;"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Cast Members
                    </span>
                  </div>
                </div>
              </li>
            </a>
            <a
              href="/genres"
              style="text-decoration: none;"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Genres
                    </span>
                  </div>
                </div>
              </li>
            </a>
            <a
              href="/videos"
              style="text-decoration: none;"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Videos
                    </span>
                  </div>
                </div>
              </li>
            </a>
          </ul>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
